hibernate-core/src/main/java/org/hibernate/internal/MultiIdentifierLoadAccessImpl.java

```diff
@@ -0,0 +1,170 @@
+		this.lockOptions = lockOptions;
+		return this;
+	}
+
+	@Override
+	public MultiIdentifierLoadAccess<T> with(CacheMode cacheMode) {
+		this.cacheMode = cacheMode;
+		return this;
+	}
+
+	@Override
+	public MultiIdentifierLoadAccess<T> with(RootGraph<T> graph, GraphSemantic semantic) {
+		this.rootGraph = (RootGraphImplementor<T>) graph;
+		this.graphSemantic = semantic;
+		return this;
+	}
+
+	@Override
+	public Integer getBatchSize() {
+		return batchSize;
+	}
+
+	@Override
+	public MultiIdentifierLoadAccess<T> withBatchSize(int batchSize) {
+		if ( batchSize < 1 ) {
+			this.batchSize = null;
+		}
+		else {
+			this.batchSize = batchSize;
+		}
+		return this;
+	}
+
+	@Override
+	public boolean isSessionCheckingEnabled() {
+		return sessionCheckingEnabled;
+	}
+
+	@Override
+	public boolean isSecondLevelCacheCheckingEnabled() {
+		return cacheMode == CacheMode.NORMAL || cacheMode == CacheMode.GET;
+	}
+
+	@Override
+	public MultiIdentifierLoadAccess<T> enableSessionCheck(boolean enabled) {
+		this.sessionCheckingEnabled = enabled;
+		return this;
+	}
+
+	@Override
+	public boolean isReturnOfDeletedEntitiesEnabled() {
+		return returnOfDeletedEntitiesEnabled;
+	}
+
+	@Override
+	public MultiIdentifierLoadAccess<T> enableReturnOfDeletedEntities(boolean enabled) {
+		this.returnOfDeletedEntitiesEnabled = enabled;
+		return this;
+	}
+
+	@Override
+	public boolean isOrderReturnEnabled() {
+		return orderedReturnEnabled;
+	}
+
+	@Override
+	public MultiIdentifierLoadAccess<T> enableOrderedReturn(boolean enabled) {
+		this.orderedReturnEnabled = enabled;
+		return this;
+	}
+
+	@Override
+	@SuppressWarnings( "unchecked" )
+	public <K> List<T> multiLoad(K... ids) {
+		return perform( () -> entityPersister.multiLoad( ids, session, this ) );
+	}
+
+	public List<T> perform(Supplier<List<T>> executor) {
+		CacheMode sessionCacheMode = session.getCacheMode();
+		boolean cacheModeChanged = false;
+		if ( cacheMode != null ) {
+			// naive check for now...
+			// todo : account for "conceptually equal"
+			if ( cacheMode != sessionCacheMode ) {
+				session.setCacheMode( cacheMode );
+				cacheModeChanged = true;
+			}
+		}
+
+		try {
+			if ( graphSemantic != null ) {
+				if ( rootGraph == null ) {
+					throw new IllegalArgumentException( "Graph semantic specified, but no RootGraph was supplied" );
+				}
+				session.getLoadQueryInfluencers().getEffectiveEntityGraph().applyGraph( rootGraph, graphSemantic );
+			}
+
+			try {
+				return executor.get();
+			}
+			finally {
+				if ( graphSemantic != null ) {
+					session.getLoadQueryInfluencers().getEffectiveEntityGraph().clear();
+				}
+			}
+		}
+		finally {
+			if ( cacheModeChanged ) {
+				// change it back
+				session.setCacheMode( sessionCacheMode );
+			}
+		}
+	}
+
+	@Override
+	@SuppressWarnings( "unchecked" )
+	public <K> List<T> multiLoad(List<K> ids) {
+		return perform( () -> entityPersister.multiLoad( ids.toArray( new Object[ 0 ] ), session, this ) );
```

Comment:
You might well reuse the `ArrayHelper.EMPTY_OBJECT_ARRAY` (which was introduced recently by https://github.com/hibernate/hibernate-orm/pull/3493) to improve performance a little bit.

Reply:
It seems we often need to invoke the #multiLoad() with Java generic collection rather than an array, maybe we can add another happy path to avoid such ugly "list -> aray" low level processing. For one thing, reference array is a deprecated usage in JDK in terms of both performance and code readability. It would be great to see the array usage as little as possible (only for primitive value array) rather than adhering to the outdated pattern.

Reply:
Out of pure curiosity, I did some JMH experiments as following:

1. a big List of String has been created in @Setup method;
2. I tested the following two dimensions:  a) whether to transform to Array first; b) whether to user iterator or index forloop

Below is the interesting results:

List + iterator: 
Benchmark              Mode  Cnt  Score   Error  Units
Benchmark.iterateList  avgt    5  0.075 ± 0.009  ms/op

List + forloop:
Benchmark              Mode  Cnt  Score   Error  Units
Benchmark.iterateList  avgt    5  0.068 ± 0.003  ms/op

Array + iterator:
Benchmark              Mode  Cnt  Score   Error  Units
Benchmark.iterateList  avgt    5  0.074 ± 0.002  ms/op

Array + forloop:
Benchmark              Mode  Cnt  Score   Error  Units
Benchmark.iterateList  avgt    5  0.074 ± 0.003  ms/op

The JDK is openjdk8. I draw some primitive conclusion as following:

- there is not much performance difference to iterate array or index forloop;
- index forloop is 10% faster than iterator in case of List;
- there is not much performance difference between List or Array as long as List usage is restricted to iterator

So suppose `list` is a List of String, the fastest way to iterate its elements is as following (10% faster than all the other three possibilities):
```
for (int i = 0; i < list.size(); i++) {
    String elem = list.get(i);
    // do something with elem
}
```

I also compared the following three ways to transform a List into an Array (in JDK8):

- list.toArray(new String[0])
- list.toArray(new String[list.size()])
- list.stream().toArray(String[]::new)

The first approach is significantly faster than the other two; whereas there is not much difference between the other two.